### DISCUSSION TITLE
A crazy way of training a model (discussion and potential implimentation)

### DISCUSSION THREAD

**Discussion by king-dahmanus** [coqui-ai/TTS#1981] [General]
Hey guys. I just got this weird idea, a new way of training a model from the cli. The following line will speak for me.
tts-trainer --model_name vits --dataset_name ljspeech --dataset_path path/to/dataset --config_path path/to/config.json --run_name "my trusty new little tts model" --run_path path/to/run/includes/checkpoints/and/samples/and/all/needed/stuff/it/is/a/folder
this has to be done after everything is prepared like the optimal process settings and the config.json.
Then we could continue the session where we left off, oh I forgot to mention implementing a logging function that saves all the settings related to the run, so when we want to finish where we left off we could do:
tts-trainer --continue "my trusty little new tts model" --run_path path/to/the/run, it will read that settings log file and searches for the latest checkpoint and continues from their.
This will make it easier for people like me to train new little things. Hopefully it won't be hard for @erogol  and the gang?

**Comment by king-dahmanus** [comment]
if I knew how to code ml stuff and edit coqui to make those changes I'd call the fork "leapfrog"

**Comment by p0p4k** [comment]
Very crazy indeed.

**Comment by king-dahmanus** [comment]
is it doable, by any chance?

**Reply by freezerain** [reply]
Am I missing something, isnt it just 
```
python ./train.py 
```
with a bunch of optional arguments that we already/usually defining in train.py.

Its like already exists but with additional level of abstraction, you can hardcode variables like paths in train.py or pass them as args.

Obviously you will have to create like a switch/searchByName function to parse model name and apropiate config template, but the more I think about it the more complexity it will need. Some models will require some additonal/different variables to be defined.
I feel like you asking for end-user accessibility for a developer tool, but maybe I dont understand use cases or minimal enough implementation scope and this is actually an QoL feature. 

Can you please explain to newbie as I am - what problem exactly you want to solve here?

**Comment by king-dahmanus** [comment]
my way is a little easier, since you don't have to do restore-prath each time since the logger can do it for you automatically

**Comment by king-dahmanus** [comment]
oh sorry didn't read the full message at first since I replied straight from reading it by email then replied here.
So like after installing the new version and setting the dataset up for training, you do that command, and it will log stuf in the dataset as I mentioned earlier, along with adding the new paths for the dataset, the config etc in a local file that coqui tts searches in when you call continue <run_name>, so you won't have to explicitely type the paths each time, especially if your training keeps crashing and you don't know what to do, as in my case, (4 gb of ram, no gpu etc). I want to be able to experiment with things with my current crappy set up. Even if it takes a month to do it's better than not being able to do it at all

**Comment by king-dahmanus** [comment]
also I don't think I remember having to do python train.py --args, but I remember the part about putting the dataset in the recipe's directory then running the python file from the recipe, is that an inefitiant way by any chance? Aside from the low specs of the craptop I mean

**Comment by erogol** [comment]
I think you can already do most of it using `train_tts.py`. What is the diff?

**Comment by king-dahmanus** [comment]
ok this is weird, you aren't understanding me at all.
so off topic, @erogol  what about the speed improvements? Y'know, the part where it automatically initiates a single synthesizer class and keeps using it automatically?  This functionality should be included in tts command as well as tts-server, for the most optimal results? Hopefully this is an easier thing to fix

**Comment by king-dahmanus** [comment]
like for the fokes who only want to test existing models as well as trainers

**Comment by king-dahmanus** [comment]
sorry if my way of speech isn't sufissiant enough, guys.  I keep suggesting things and most of the time no one answers my questions and enquiries, not exclusively in coqui's chatting platforms either, so I start to temper and stuff, so, apologies once more

**Reply by p0p4k** [reply]
The thing is, you are suggesting ideas without the actual effort of trying to implement stuff yourself. Try submitting a PR with your ideas in it and let the reviewers change/correct/access it. That way, you will learn better, have your ideas implemented and not get mad at yourself while yelling at a wall. Plus this is a solid community that appreciates even the slightest efforts that you will make in your contribution, so do not be shy to try out things yourself first. Thanks and good luck! :-)

**Comment by king-dahmanus** [comment]
hey @p0p4k 
Thanks for your suggestions, but currently my level of knowledge doesn't allow me to make things myself, which is why I keep suggesting verbal things. When I suggest a feature related to the interface, some people won't agree with me which is acceptable, so it doesn't get implemented.
If I ask about certain bugs, I get no answers what so ever, like the tts speed issue. My usecases require instant synthesis with a maximum delay of 20 ms. Since I can't test things optimally due to my low specs, I won't be able to contribute based on *modern* setups, so my idea doesn't get serious attention.
If I try to suggest revolutionizing arabic tts, even if I provide *untested theories*, no one will be able to actually experiment with arabic, as I''ve noticed a *practically 0* amount of arabic related activity in this project. And since you guys won't be able to experiment with arabic since you very likely don't know arabic, my ideas won't get implemented. This is what I'm facing. I like this project and I'd contribute if I could. If anything, the amount of threads I post here represent my huge interest in the project :)
Thanks for reading, and hope to read your replies soon Those were some examples on what I would likely do if I had the knowledge, time, and resources to do.

**Reply by p0p4k** [reply]
@king-dahmanus You basically answered your own question about why no one replies or takes you seriously yet. I hope you acquire the required coding skills and bounce back stronger. Good luck.

**Comment by king-dahmanus** [comment]
hey thanks p0p4k. So after the python basics, what do I need to learn to be able to contribute stuff to coqui? Like the big things I mentioned, maybe some small stuff in the beginning

**Reply by p0p4k** [reply]
Yes,
1) learn python basics first, 
2) watch other beginner projects on YouTube, 
3) read other people's code, understand what it does
4) try to aim a create small project (maybe like 'making TTS synth easier for newbs'), 
5) Google about what needs to be done to get there.
It will not be a small sprint, but a marathon and once you reach there, come back and contribute. 
**Do not let anyone tell you that you cannot do it. Talk less, work more.**
Good Luck!

**Comment by king-dahmanus** [comment]
Thanks! So not all youtube videos work for me since I'm blind and all, and a lot of the channels just go "Enter this code here and click in the right corner with your mouse and drag this here untill it looks good and then ...) and I can't understand any of that obviously. So what beginner projects do you recommend that are audibly documented in a way that *everyone* can understand?

**Reply by p0p4k** [reply]
I really do not have any clue about this edge case. In this case maybe buying physical coding books with braille support  would be an option, I guess?

**Comment by king-dahmanus** [comment]
no no no, not braille. I'm requesting tutorials where the teacher actually speaks the code they're typing. God, the media of the world aren't doing a good job of teaching people about how blind people use their devices, so most people believe tthat blind people can't do anything on their own.

**Reply by p0p4k** [reply]
Oops. Forgive my ignorance on this subject. I really do not know anything in this regard.

**Comment by king-dahmanus** [comment]
don't worry about it, no one knows everything :)
so let me know if you find a chanel with blind friendly content, k?

**Reply by p0p4k** [reply]
Sure thing! Will keep an eye for such stuff from now on.

**Comment by king-dahmanus** [comment]
Thank you! I already have the basics part covered, I just need the part
about projects I can exercise on.

**Reply by Ca-ressemble-a-du-fake** [reply]
@king-dahmanus I found this tutorial interesting and the teacher explains what they type in.  https://youtu.be/DooxDIRAkPA

**Comment by king-dahmanus** [comment]
ey thanks cara. :)
